lib/puppet/parser/environment_compiler.rb

```diff
@@ -40,20 +46,62 @@ def compile
     end
   end
 
-  def add_resource(scope, resource)
-    type = resource.resource_type
-    # At topscope, only applications and Class[main] are allowed. Elsewhere,
-    # only components and capabilities are allowed.
-    if scope == @topscope
-      if ! (type.is_a?(Puppet::Resource::Type) && type.application?) && resource != @main_resource
-        raise ArgumentError, "Only applications are allowed at topscope, not #{resource.ref}"
-      end
+  # Prunes the catalog by dropping all resources that are not contained under the Site (if a site expression is used).
+  # As a consequence all edges to/from dropped resources are also dropped.
+  # Once the pruning is performed, this compiler returns the pruned list when calling the #resources method.
+  # The pruning does not alter the order of resources in the resources list.
+  #
+  def prune_catalog
+    # Everything under Class[main], that is not under (inclusive of) Site[site] should be pruned as those resources
+    # are intended for nodes in a node catalog.
+    #
+    the_main_class_resource = @catalog.resource('Class', '')
+    the_site_resource = @catalog.resource('Site', 'site')
+
+    # Get downstream vertexes returns a hash where the keys are the resources and values nesting level
+    rooted_in_main = @catalog.downstream_from_vertex(the_main_class_resource).keys
+    if the_site_resource
+      keep_from_site = @catalog.downstream_from_vertex(the_site_resource).keys
+      keep_from_site << the_site_resource
     else
-      unless (resource.is_application_component? || resource.is_capability?)
```

Comment:
Yes, that seems right to me; inside the `site` statement we want, after evaluation, applications and their components. I am not entirely clear whether containment is set up right, but assuming it is, we should have inside `site` something like `app_instance -> component -> other_stuff` - we want the app_instance in there and the component; the other_stuff is not an error but we don't care about it. I am not sure whether that happens, but during environment compilation it's sufficient to evaluate just to the point where we know the type/title of components and the type/title of capability resources referenced by `export`, `consume` or any of the standard relationship metaparams. IOW, we do not need to actually evaluate the bodies of components.